clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1074,8 +1074,17 @@ static bool buildArgTys(ASTContext &Context, CXXRecordDecl *KernelObj,
           cast<CXXRecordDecl>(ArgTy->getAs<RecordType>()->getDecl());
       if (!RD->hasTrivialCopyConstructor()) {
         Context.getDiagnostics().Report(
-            Fld->getLocation(), diag::err_sycl_non_trivially_copyable_type)
-            << ArgTy;
+            Fld->getLocation(),
+            diag::err_sycl_non_trivially_copy_ctor_dtor_type)
+            << 0 << ArgTy;
+        AllArgsAreValid = false;
+        continue;
```

Comment:
Let's assume I use a class with both: non-trivial constructor and destructor. What kind of diagnostic will I get?

Another case: I enable SYCL-1.2.1 mode (i.e. enable standard layout restirction) for the previous test case and it's non-standard layout on top of that.

Basically what I'm hinting at is "do we need continue statements in all these branches?".

Reply:
You'll only get the 1st one.  That is pretty typical in the CFE to only diagnose 1 problem with a thing at a time.

Reply:
I think we can provide results of all three checks at once if we move them to a separate function and skip argument lowering if this function return code is not successful.
E.g.
```c++
bool checkStructureKernelArg(...) {
      bool IsValid = true;

      if (Context.getLangOpts().SYCLStdLayoutKernelParams) {
        if (!ArgTy->isStandardLayoutType()) {
          Context.getDiagnostics().Report(Fld->getLocation(),
                                          diag::err_sycl_non_std_layout_type)
              << ArgTy;
          IsValid = false;
        }
      }

      CXXRecordDecl *RD =
          cast<CXXRecordDecl>(ArgTy->getAs<RecordType>()->getDecl());
      if (!RD->hasTrivialCopyConstructor()) {
        Context.getDiagnostics().Report(
            Fld->getLocation(),
            diag::err_sycl_non_trivially_copy_ctor_dtor_type)
            << 0 << ArgTy;
          IsValid = false;
      }
      if (!RD->hasTrivialDestructor()) {
        Context.getDiagnostics().Report(
            Fld->getLocation(),
            diag::err_sycl_non_trivially_copy_ctor_dtor_type)
            << 1 << ArgTy;
          IsValid = false;
      }
      return IsValid;
}

...
  if (!checkStructureKernelArg(...)) {
    AllArgsAreValid = false;
    continue;    
  }
```

Reply:
Ah! Apparently my post yesterday didn't actually submit...

We COULD do that, and I will if you wish for me to, but having multiple errors on the same semantic location is highly irregular in the frontend.  We nearly always give up on something immediately after the first error.  So I can do that if thats what you want, but it would be fairly novel for the frontend.

Reply:
I'm okay to leave it as is.
We will have a chance to check the approach with the clang community when this patch is uploaded to Phabricator.